janus_collector/src/lib.rs

```diff
@@ -236,16 +236,75 @@ impl<P> CollectJob<P> {
 
 #[derive(Derivative)]
 #[derivative(Debug)]
-/// The result of a collect request poll operation. This will either provide the aggregate result
+/// The result of a collect request poll operation. This will either provide the collection result
 /// or indicate that the collection is still being processed.
-enum PollResult<T> {
-    /// The aggregate result from a completed collect request.
-    AggregateResult(#[derivative(Debug = "ignore")] T),
+enum PollResult<V: vdaf::Collector>
```

Comment:
I prefer keeping the type constraints "pushed down" here, and parametrizing `PollResult` and `CollectionResult` on the aggregate result type, not the VDAF type. This models what they do better, as they only act as a container for this result, and don't otherwise make use of the `Vdaf` trait. It also would allow eliminating many where clauses when working with just these two types.

Reply:
Hmm, I made this change, but IMO it's clearer to talk about a `PollResult`/`CollectionResult` for a given VDAF -- otherwise, the type system doesn't make it clear which VDAF the concrete result type is associated with. That is, specifying the VDAF (& using the VDAF's associated type to specify the concrete stored type) over specifying the concrete type directly provides more information to both the type system & the programmer, and I think there are no cases where we can't specify the VDAF in use.

The infectious `where` clauses are indeed a pain. I took some time looking to see if an equivalent clause could be specified in a non-infectious way, but as far as I could tell this is a (particularly annoying) rough edge of Rust at the moment.

Reply:
I wonder if we could spare ourselves some `where`s in Janus if we defined a supertrait over `prio::vdaf::Vdaf` that is just `Vdaf` plus the extra bounds we end up writing everywhere. Then we could make everything generic over `JanusVdaf` or `JanusAggregator` instead of `prio::vdaf::Vdaf` or `prio::vdaf::Aggregator`. I wouldn't tackle that here, though.